lib/include/public/ctmacros.hpp

```diff
@@ -127,4 +127,24 @@
 #define EVTSDK_LIBABI_CDECL MATSDK_LIBABI_CDECL
 #define EVTSDK_SPEC         MATSDK_SPEC
 
+/* Implement struct packing for stable FFI C API */
+#ifdef __clang__
+# define MATSDK_PACKED_STRUCT __attribute__((packed))
```

Comment:
This has the side effect of setting struct alignment to 1 which is probably not intended here? https://gcc.godbolt.org/z/jMK4xzbK8

The structs seem like they have been laid out so that natural alignment works, so I would suggest removing these macros unless there's a known issue. static_assert on size/offsets may give some safety?

Reply:
@stephenatwork - there is (was) a known issue, at least on Mac. Padding depends on compiler optimization level. Current code has been tested on Quest (ARM64) and Win-x64, but I agree that adding static asserts for checking the offset and size totally makes sense.

Test app performs ingestion via C API, reconstructing into C++ EventProperties, then performs for-each - compare that there's a match. I can add a few props that'd be of size larger than 256 bytes / 65536 bytes, etc. to the test as well.

Let me add static asserts for size and offset checks.

Reply:
> Padding depends on compiler optimization level
🤯
OK if there are weird compilers to deal with then adding an explicit alignment attribute can avoid all kinds of awful shuffling (especially on platforms like ARM where unaligned loads are bad). So maybe `__attribute__((packed,align(4)))` would do it?

Reply:
> (especially on platforms like ARM where unaligned loads are bad)

You'd actually fit more props in 64 byte data cache line with packing. Most loads and stores for strings - are pointers to aligned memory prepared by C# anyways:

- Perf is better because you can fit more props (e.g. 8 numbers) in [one data cache line](https://developer.arm.com/documentation/ddi0500/e/level-1-memory-system/about-the-l1-memory-system#:~:text=Data%20side%20cache%20line%20length%20of%2064%20bytes.)

- Perf is unaffected for string props: since once you got a pointer to string - you do loads and stores on aligned memory. C# calling into C on Win-x64 and Oculus would store the actual strings aligned.

I can address a potential issue for 32-bit. C# code already makes an assumption about the packed struct layout. I can add padding for 32-bit, so that the layout and sizing is identical. I'll push a change that adds a static assert check for padding/sizes.

Reply:
Regarding Pii tags on individual props:
- This used to be a cool feature supported by the old Kusto path, with values obfuscated before being exported to DB.
- We moved on to a different routing model where this old feature is deprecated. There is an alternate technique to tag the whole events table as sensitive.

Reply:
I kept the order for the following reasons:

- This already works for us on ARM64 and Win-x64, and we don't need more.

- Same order of fields as for older C API / ABI (though different padding/alignment). This allows to easily `#ifdef` and fallback to old C API. Trying to appease the customers who for some reason need to retain the older `v3.1` C ABI struct layout.

- Added an extra TC that verifies padding, alignment, struct size are consistent on 32-bit and 64-bit. Added support for Win32 32-bit / x86 (just in case, although my team doesn't require it). One IL assembly can now load either 32-bit or 64-bit version of the Client Telemetry shared lib, with both flavors exposing the same C API, basically with the same struct ABI across different platforms and compilers. Pointers to `const char *` and `void *`, of course, remain platform-dependent. But 32-bit version gets 32-bit padding in between, whereas 64-bit version occupies the full 64-bit for the pointers with no padding.

- Unaligned read problem is superficial: it won't actually occur on Intel and ARM. It does not affect perf either - because you fit more in fewer data cache lines. When either C/C++ or C# allocs an array of `evt_prop` structs on their heap, structs are `64(name)+32(type)+64(value)+32(piiKind)`-sized, packed back-to-back. More of them fits that way in a smaller-sized contiguous region of memory. Pointer to that region itself is aligned. Pointers to the actual props, e.g. strings that get allocated separately - are also all aligned. IF we ever need to support `RISC` with this... Well, there is no C# support for that. If someone would like to continue using SDK on `RISC`, then for those customers I added an option to revert back to original struct definition with a compiler define `HAVE_MAT_ABI_V3_1_0` . That way the padding decisions are left to compiler imagination. No C# support in that case though.

Given that it works on Android ARM64 and Win-x64 (for our org in the Mesh apps). And none of the CI checks of this PR stumbled upon the struct layout, including Apple mobile device tests and Android - all passing. Plus size-checks - all passed, I think we good.

Reply:
SGTM. Back compat plan is nice to have too.

On unaligned loads, maybe all platforms we care about handle them OK now. But have a look at
https://gcc.godbolt.org/z/6jWY6EP76 The runtime alignment doesn't matter so much - the compiler must emit code for the least aligned version it knows about. For some platforms it's only a little (or no) extra code. For some it can be a lot! (Gamedev made me super sensitive to code size)

Reply:
Sounds good. Resolving this conversation. We can revisit later, if ever needed be.